### DISCUSSION TITLE
Received incompatible tensor with shape (127,) when attempting to restore variable with shape (124,) and name dense/bias:0.

### DISCUSSION THREAD

**Discussion by suresrikanth101** [mindee/doctr#1407] [Q&A]
### Bug description

I tried to finetune recognization with tensorflow framework in my labelled data, when I load weights,Iam getting error like ValueError: Received incompatible tensor with shape (127,) when attempting to restore variable with shape (124,) and name dense/bias:0. But checkpoint, weights.data and weights.index files are stored properly. plz help 

### Code snippet to reproduce the bug

from doctr.models import ocr_predictor, db_resnet50, crnn_vgg16_bn
# Load custom recognition model
reco_model = crnn_vgg16_bn(pretrained=False, pretrained_backbone=False)
reco_model.load_weights(r"C:\Users\SSure\OneDrive - Genzeon Corporation\Desktop\POc\doctr\crnn_vgg16_bn_20231211-180856/weights")
predictor = ocr_predictor(det_arch="linknet_resnet18", reco_arch=reco_model, pretrained=True)

### Error traceback

ValueError: Received incompatible tensor with shape (127,) when attempting to restore variable with shape (124,) and name dense/bias:0.

### Environment

DocTR version: 1.9.0
TensorFlow version: 2.15.0
PyTorch version: 2.1.1+cpu (torchvision 0.16.1+cpu)
OpenCV version: 4.5.4
OS: Could not collect
Python version: 3.10.11
Is CUDA available (TensorFlow): No
Is CUDA available (PyTorch): No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

### Deep Learning backend

is_tf_available: True
is_torch_available: False

**Comment by felixT2K** [comment]
Hi @suresrikanth 👋🏼 ,

you missed to specify the vocab the model was trained on :)

See for example:

```
from doctr.models import ocr_predictor, crnn_vgg16_bn
from doctr.datasets import VOCABS

reco_model = crnn_vgg16_bn(pretrained=False, pretrained_backbone=False, vocab=VOCABS["german"])  <--- this
reco_model.load_weights("<path_to_checkpoint>/weights")

predictor = ocr_predictor(det_arch='linknet_resnet18', reco_arch=reco_model, pretrained=True)
```

**Comment by suresrikanth101** [comment]
Hi @felixT2K,
I was trying to increase my accuracy on handwritten text, so I prepared handwritten labeled data and started finetuning with that data. I was getting error like above. Note: Here I used for english only, in this case no need to place vocab at that place, am I right?   and   Especially I did labels on doctor notes and handwritten text in my documents, what should I provide in vocabs. Any custom key value would add in the vocab.py file

**Comment by felixT2K** [comment]
Can you please try the following:

```
reco_model = crnn_vgg16_bn(pretrained=False, pretrained_backbone=False, vocab=VOCABS["french"])
reco_model.load_weights("<path_to_checkpoint>/weights")
```

The problem is that the default from crnn_vgg16_bn is still legacy_french with 124 characters (it was one of the first trained models) but the default in the training script is french with 127 characters

**Comment by felixT2K** [comment]
Could you please share the code snippet you used with loading your custom model ? 
changing the recognition model has no impact on the detection model so this looks super weird 😅 

For example the detection for:

```
reco_model = crnn_vgg16_bn(pretrained=False, pretrained_backbone=False)
reco_model.load_weights("<path_to_checkpoint>/weights")
predictor = ocr_predictor(det_arch="db_resnet50", reco_arch=reco_model, pretrained=True)
```

and

```
predictor = ocr_predictor(det_arch="db_resnet50", reco_arch="crnn_vgg16_bn", pretrained=True)
```

should be the same

**Comment by suresrikanth101** [comment]
I used below code
reco_model = crnn_vgg16_bn(pretrained=False, pretrained_backbone=False, vocab=VOCABS["french"])
reco_model.load_weights("<path_to_checkpoint>/weights")
predictor = ocr_predictor(det_arch="linknet_resnet18", reco_arch=reco_model, pretrained=True)
doc = DocumentFile.from_images(r'<image path>')
result = predictor(doc)

**Comment by felixT2K** [comment]
A now i got it you used 2 different detection models .. in this case you should use `db_resnet50` for detection if the results are better instead of `linknet_resnet18`

**Comment by suresrikanth101** [comment]
yes you are right, just now i changed to 'db_resnet50', it is working as expected. Thank you .

**Comment by felixT2K** [comment]
So sounds like the issue is fixed on your side ? :)

**Comment by suresrikanth101** [comment]
Hi @felixT2K,
When I am taking inferences, text inside the document detecting properly, but when it comes to recognization it fails, in the json response value is empty space and it also  provided corresponding boundary boxes as well. plz find below screenshot of json response.
<img width="660" alt="2" src="https://github.com/mindee/doctr/assets/68548376/370d4509-6397-4661-bf74-62ebae093c8b">

I have doubt, the updated weights are a combination of the original pre-trained weights and the new weights learned from My data. This means that the model still retains its general knowledge of character recognition, but it also has additional knowledge that is specific to your data and can help it perform better on similar tasks. Am I correct?

**Comment by felixT2K** [comment]
H @suresrikanth,

This depends on how you have trained the model, have you used `--pretraned` and fine tuned the model or trained it from scratch ?
How many samples are in your dataset ?
Also `crnn_vgg16_bn` is not the optimal choice for handwritten recognition you should use a model which uses attention (transformer) i suggest `parseq` because it is small and really powerful

**Reply by suresrikanth101** [reply]
actually trained recognization from scratch, I want to use pretrained weights in addition to that include custom weights.
total samples were 250. 
Has doctr ocr offer direct option for combining pre-trained weights with custom model weights.
Help me where exactly i can change

**Reply by felixT2K** [reply]
@suresrikanth you can pass `--pretrained` and `--vocabs=french` to the training script

**Reply by suresrikanth101** [reply]
I used this below code 
python references/recognition/train_tensorflow.py crnn_vgg16_bn --train_path "C:\Users\POc\docTR-trainingDataset" --val_path "C:\Users\POc\doctr-valDataset"  --epochs 5  --pretrained --vocab french . getting error 
ValueError: Received incompatible tensor with shape (124,) when attempting to restore variable with shape (127,) and name dense/bias:0.

**Reply by felixT2K** [reply]
Hey @suresrikanth 
Yeah i remember the issue could you please check any other model with the same config ?
For example `crnn_mobilenet_v3_small` or `vitstr_small`

The issue exists only with TensorFlow as backend and it's known.
Because we don't have logic to manipulate or ignore layer shape mismatches which in this case is raised by the last classifier layer.
The model was trained on `legay_french` which contains 124 chars and you try to train on a vocab which contains 127 chars 
`crnn_vgg16_bn` is the only model which is trained on legacy_french all other models are pretrained on the normal french vocab so `crnn_vgg16_bn` should be the only model which raises this issue for your dataset / config

**Reply by felixT2K** [reply]
We can fix this issue hopefully after the next TF release. 
Where will be forced to refactor from keras 2 to keras 3 (keras core).

**Comment by felixT2K** [comment]
The ground issue seems to be solved i will covert it to a discussion :)
